drivers: i3c: cdns i3c driver

XenuIsWatching opened the pull request:
This is initial support of the i3c cdns controller.
This also fixes few issues that were found during development. Please view their commit message for details.

XenuIsWatching committed:
i3c: rename is_primary to is_secondary

Rename is_primary to is_secondary. The justification for this is
because it is less likely to have something configured to be
secondary, and the 0 value would be if it is primary.

Signed-off-by: Ryan McClelland <ryanmcclelland@meta.com>

XenuIsWatching committed:
i3c: fix macro generation with multiple devices

When generating an array of devices with I3C_DEVICE_ARRAY_DT_INST
and I3C_I2C_DEVICE_ARRAY_DT_INST when the number of devices is
greater than 1, a comma was not inserted between each struct with
macro. This would give an error preventing compiling.

Signed-off-by: Ryan McClelland <ryanmcclelland@meta.com>

XenuIsWatching committed:
i3c: add write to tx fifo target api

Some controllers will "automatically" NACK any read request to them
if there is no data in the TX FIFO. This would prevent use of the callback
function read_requested_cb which is expected to be called from an ISR.
i3c_target_tx_write was added to give applications a direct way to load
data in to the tx fifo.

Signed-off-by: Ryan McClelland <ryanmcclelland@meta.com>

XenuIsWatching committed:
i3c: add NBCH header flag for private transfers

According to section 5.1.2.2.2 of the I3C Specification v1.1.1. A
controller may skip the 0x7E broadcast header for private
transfers as an optimization if it is know that no I3C Target on
the bus has a spike filter.

Signed-off-by: Ryan McClelland <ryanmcclelland@meta.com>

XenuIsWatching committed:
i3c: GETMWL and GETMRL may be optionally supported if no settable limit

According to section 5.1.9.3.5 and 5.1.9.3.6 of the I3C Specification
v1.1.1. This CCC is may be optionally supported if the target device
has no settable limit.

Signed-off-by: Ryan McClelland <ryanmcclelland@meta.com>

XenuIsWatching commented on drivers/i3c/Kconfig.cdns:
```diff
@@ -0,0 +1,16 @@
+config I3C_CADENCE
+	bool "Cadence I3C driver"
+	select I3C_IBI_WORKQUEUE if I3C_USE_IBI
+	default y
```
fixed

XenuIsWatching commented on drivers/i3c/Kconfig.cdns:
```diff
@@ -0,0 +1,16 @@
+#
+# SPDX-License-Identifier: Apache-2.0
+
+DT_COMPAT_CADENCE_I3C := cdns,i3c
```
fixed

XenuIsWatching commented on drivers/i3c/i3c_cdns.c:
```diff
@@ -0,0 +1,2459 @@
+		if (cdns_i3c_rx_fifo_empty(config)) {
+			return -EIO;
+		}
+		val = sys_cpu_to_le32(sys_read32(config->base + RX_FIFO));
```
fixed

XenuIsWatching commented on drivers/i3c/i3c_cdns.c:
```diff
@@ -0,0 +1,2459 @@
+	uint32_t dev_id_rr1;
+	uint32_t dev_id_rr2;
+
+	/* program I3C devices */
```
fixed

XenuIsWatching committed:
i3c: add cdns i3c driver

This gives initial support to the cadence i3c controller

Signed-off-by: Ryan McClelland <ryanmcclelland@meta.com>

dcpleung commented on drivers/i3c/i3c_common.c:
```diff
@@ -402,23 +402,13 @@ static int i3c_bus_setdasa(const struct device *dev,
-			desc->dynamic_addr = desc->static_addr;
+			desc->dynamic_addr = (desc->init_dynamic_addr ? desc->init_dynamic_addr
+								      : desc->static_addr);
+			i3c_reattach_i3c_device(dev, desc, desc->static_addr);
```
Need to handle when this call returns errors.

XenuIsWatching replied:
This was something that I did think about.... but I decided not to do anything if there is an error here as this api is optional and could return ENOSYS..... but there is the issue of if that address is not available.

I do think this reattach api could be made to handle the addr_list before calling in to the driver implemented function.... but that would require some refactoring of how i3c controllers have their config/data commomly defined. I wrote a bit about it in this issue https://github.com/zephyrproject-rtos/zephyr/issues/51828

XenuIsWatching commented:
I made a small change to the i3c_reattach_i3c_device removing a redundant parameter.

XenuIsWatching commented:
I found a single header that was 'auto-included' by my IDE that shouldn't been there in `i3c_cdns.c`. I've removed it and did a rebase.

dcpleung commented:
Do you have a board config that would utilize this driver? I think it's a general policy that CI must be able to build the newly introduced drivers so CI can make sure they build (and also to prevent future breakage).

XenuIsWatching commented:
> Do you have a board config that would utilize this driver? I think it's a general policy that CI must be able to build the newly introduced drivers so CI can make sure they build (and also to prevent future breakage).

No, there aren't any that I know of :( this is all intended with custom ASICs and FPGAs for those to use. While I understand your point... I'm not sure that it is general policy, but it probably should be. There ought to be something like `/tests/drivers/build_all/...` for these cases

XenuIsWatching commented:
> I found a single header that was 'auto-included' by my IDE that shouldn't been there in `i3c_cdns.c`. I've removed it and did a rebase.

Apparently, I was wrong. That header was added by "me", and was to be there

cfriedt commented:
> Do you have a board config that would utilize this driver? I think it's a general policy that CI must be able to build the newly introduced drivers so CI can make sure they build (and also to prevent future breakage).

@teburd, @dcpleung - we should probably have some entries under `tests/drivers/build_all` for i2c / i3c (build-only obviously). Likely out of scope for this PR.

dcpleung commented:
> > Do you have a board config that would utilize this driver? I think it's a general policy that CI must be able to build the newly introduced drivers so CI can make sure they build (and also to prevent future breakage).
> 
> @teburd, @dcpleung - we should probably have some entries under `tests/drivers/build_all` for i2c / i3c (build-only obviously). Likely out of scope for this PR.

I think it would be a good idea to do that for this driver. The NXP I3C driver is being built via sensor sample so that is covered.

XenuIsWatching committed:
tests: drivers: build_all: add build for i3c_cdns

The Cadence I3C has no in-tree board with it to built with. This
adds a test build-only for the cadence i3c peripheral attaching
with the qemu_cortex_m3 board as that appears to be generic
enough.

Signed-off-by: Ryan McClelland <ryanmcclelland@meta.com>

XenuIsWatching commented:
> @XenuIsWatching - for reference, I just made an fpga build all test under `tests/drivers/build_all/fpga`. Feel free to ping if you need help.

Thanks, I think I figured it out. I just created a new folder, but it builds under the `qemu_cortex_m3` board as that seems generic enough.
